service/src/main/java/bio/terra/workspace/service/workspace/flight/RemoveUserFromWorkspaceFlight.java

```diff
@@ -0,0 +1,50 @@
+package bio.terra.workspace.service.workspace.flight;
+
+import bio.terra.stairway.Flight;
+import bio.terra.stairway.FlightMap;
+import bio.terra.workspace.common.utils.FlightBeanBag;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.iam.model.WsmIamRole;
+import bio.terra.workspace.service.job.JobMapKeys;
+import java.util.UUID;
+
+public class RemoveUserFromWorkspaceFlight extends Flight {
+
+  public RemoveUserFromWorkspaceFlight(FlightMap inputParameters, Object applicationContext) {
+    super(inputParameters, applicationContext);
+
+    FlightBeanBag appContext = FlightBeanBag.getFromObject(applicationContext);
+
+    UUID workspaceId =
+        UUID.fromString(inputParameters.get(WorkspaceFlightMapKeys.WORKSPACE_ID, String.class));
+    AuthenticatedUserRequest userRequest =
+        inputParameters.get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+    String userToRemove = inputParameters.get(WorkspaceFlightMapKeys.USER_TO_REMOVE, String.class);
+    WsmIamRole roleToRemove =
+        WsmIamRole.valueOf(
+            inputParameters.get(WorkspaceFlightMapKeys.ROLE_TO_REMOVE, String.class));
+
+    // Flight plan:
+    // 0. (Pre-flight): Validate that the user is directly granted the specified workspace role.
+    //  WSM does not manage groups, so users with indirect group-based access cannot be removed
```

Comment:
How would a user have gotten such indirect access in the first place?

Reply:
WSM supports granting roles to Sam managed groups as well as users. So, `my-research-group@sam-group-domain.com` can be a reader of my workspace, and you would have access via group membership. In that case, WSM won't know when that group's membership changes, though it could still remove the entire group from the workspace.

Reply:
Is this one of the use cases for the Pub/Sub queue?

Reply:
If you mean getting notification of access changes from Sam via PubSub, yes, this is one of the use cases for that.